netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -423,7 +841,478 @@ net_ebpf_extension_sock_addr_authorize_connection_classify(
+    ASSERT((compartment_id == UNSPECIFIED_COMPARTMENT_ID) || (compartment_id == sock_addr_ctx.compartment_id));
+    if (compartment_id != UNSPECIFIED_COMPARTMENT_ID && compartment_id != sock_addr_ctx.compartment_id) {
+        // The client is not interested in this compartment Id.
+        NET_EBPF_EXT_LOG_MESSAGE_UINT32(
+            NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
+            NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+            "The cgroup_sock_addr eBPF program is not interested in this compartment ID",
+            sock_addr_ctx.compartment_id);
+
+        verdict = BPF_SOCK_ADDR_VERDICT_PROCEED;
+        goto Exit;
+    }
+
+    // Get the connection context for this connection.
+    status = _net_ebpf_ext_get_connection_context(
+        incoming_metadata_values->transportEndpointHandle, &sock_addr_ctx, &connection_context);
+    if (!NT_SUCCESS(status)) {
+        // We did not find any connection context for this AUTH request. Block.
+        verdict = BPF_SOCK_ADDR_VERDICT_REJECT;
+        goto Exit;
+    }
+    verdict = connection_context->value.verdict;
+
+    _net_ebpf_ext_delete_connection_context(connection_context);
+    connection_context = NULL;
+
+Exit:
+    classify_output->actionType = (verdict == BPF_SOCK_ADDR_VERDICT_PROCEED) ? FWP_ACTION_PERMIT : FWP_ACTION_BLOCK;
+    // Clear FWPS_RIGHT_ACTION_WRITE only when it is a hard block.
+    if (classify_output->actionType == FWP_ACTION_BLOCK) {
+        classify_output->rights &= ~FWPS_RIGHT_ACTION_WRITE;
+    }
+
+    NET_EBPF_EXT_LOG_MESSAGE_UINT64_UINT64_UINT64(
+        NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
+        NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+        "authorize_connection_classify",
+        incoming_metadata_values->transportEndpointHandle,
+        sock_addr_ctx.protocol,
+        verdict);
+
+    return;
+}
+
+static NTSTATUS
+_net_ebpf_ext_process_redirect_verdict(
+    uint32_t verdict,
+    _In_ const bpf_sock_addr_t* original_context,
+    _In_ const bpf_sock_addr_t* redirected_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t classify_handle,
+    HANDLE redirect_handle,
+    _Out_ bool* redirected,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    FWPS_CONNECT_REQUEST* connect_request = NULL;
+    bool commit_layer_data = false;
+
+    *redirected = false;
+
+    if (verdict == BPF_SOCK_ADDR_VERDICT_PROCEED) {
+        // Check if destination IP and/or port have been modified.
+        bool address_changed = !_net_ebpf_ext_compare_destination_address(redirected_context, original_context);
+        if (redirected_context->user_port != original_context->user_port || address_changed) {
+            *redirected = true;
+
+            status = FwpsAcquireWritableLayerDataPointer(
+                classify_handle, filter->filterId, 0, (PVOID*)&connect_request, classify_output);
+            if (!NT_SUCCESS(status)) {
+                NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE_UINT64_UINT64(
```

Comment:
Use BAIL_ON_API_FAILURE style of macro, that does the error checking and goto itself.